### ISSUE TITLE
[Bug] Kimi-K2.6 (W4A16 INT4, compressed-tensors) produces deterministic single-token repetition loops on greedy decode on Blackwell GB200 (SM 100 / aarch64 / cu130); identical setup works correctly on H100

### ISSUE THREAD

**Issue by songwangnlp** [sgl-project/sglang#24091]
### Summary                                                                                                                                                                                                                                                                                         
   
  Serving `moonshotai/Kimi-K2.6` (W4A16 INT4 / `compressed-tensors` / `pack-quantized`) on a 2-node × 4-GPU GB200 cluster (Blackwell SM 100, aarch64, cu130) under sglang 0.5.10.post1 collapses into a single-token (or short-cycle) repetition loop on greedy decode (`temperature=0`). The same    
  model weights served on H100 (Hopper SM 90, x86_64, cu128) at the same sglang version produce clean, fluent output on the same prompt.
                                                                                                                                                                                                                                                                                                      
  The bug appears to be in the W4A16 INT4 GEMM kernel (Marlin) compiled into `sgl_kernel` for `cu130 / aarch64 / SM 100`. It is below the orchestration layer: changing `--moe-runner-backend` or disabling CUDA graphs changes which token the model gets stuck on, but does not fix the corruption.
                                                                                                                                                                                                                                                                                                      
  ### Environment
                                                                                                                                                                                                                                                                                                      
  |  | Working (reference) | Broken |                                                                                                                                                                                                                                                                 
  |---|---|---|
  | Hardware | 16× H100 (Hopper SM 90), 2 nodes × 8 | 8× GB200 (Blackwell SM 100), 2 nodes × 4 |                                                                                                                                                                                                      
  | Topology | TP=8 PP=2 | TP=4 PP=2 |                                                                                                                                                                                                                                                                
  | sglang | 0.5.10.post1 | 0.5.10.post1 |
  | sgl_kernel | 0.4.1 (cu129 / x86_64 wheel) | **0.4.1 AND 0.4.1.post1 both broken** (cu130 / aarch64 wheel) |                                                                                                                                                                                       
  | PyTorch | 2.9.1+cu128 | 2.9.1+cu130 |                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                      
  Model `quantization_config` (from `config.json`):                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                      
  ```json                                                                                                                                                                                                                                                                                             
  {                                                               
    "quant_method": "compressed-tensors",
    "format": "pack-quantized",
    "quantization_status": "compressed",
    "config_groups": {
      "group_0": {
        "weights": {                                                                                                                                                                                                                                                                                  
          "num_bits": 4,
          "group_size": 32,                                                                                                                                                                                                                                                                           
          "strategy": "group",                                    
          "symmetric": true,
          "type": "int"
        }
      }
    },                                                                                                                                                                                                                                                                                                
    "ignore": [
      "lm_head",                                                                                                                                                                                                                                                                                      
      "re:.*self_attn.*",                                         
      "re:.*shared_experts.*",
      "re:.*mlp\\.(gate|up|gate_up|down)_proj.*"
    ]                                                                                                                                                                                                                                                                                                 
  }
  ```                                                                                                                                                                                                                                                                                                 
                                                                  
  ### Reproduction

  Head launch (worker identical with `--node-rank 1`):                                                                                                                                                                                                                                                
   
  ```bash                                                                                                                                                                                                                                                                                             
  python -m sglang.launch_server \                                
    --host 0.0.0.0 --port 9090 \                                                                                                                                                                                                                                                                      
    --model-path /path/to/Kimi-K2.6 \
    --trust-remote-code \                                                                                                                                                                                                                                                                             
    --served-model-name Kimi-K2.6 \                               
    --quantization compressed-tensors \                                                                                                                                                                                                                                                               
    --tensor-parallel-size 4 --pipeline-parallel-size 2 \         
    --nnodes 2 --node-rank 0 \                                                                                                                                                                                                                                                                        
    --dist-init-addr <MASTER_IP>:25000 --nccl-port 25001 \                                                                                                                                                                                                                                            
    --tool-call-parser kimi_k2 --reasoning-parser kimi_k2 \                                                                                                                                                                                                                                           
    --mem-fraction-static 0.85 \                                                                                                                                                                                                                                                                      
    --max-total-tokens 128000 --context-length 128000 \                                                                                                                                                                                                                                               
    --attention-backend flashinfer \                                                                                                                                                                                                                                                                  
    --moe-runner-backend cutlass \                                
    --kv-cache-dtype bf16                                                                                                                                                                                                                                                                             
  ```                                                             
                                                                                                                                                                                                                                                                                                      
  Deterministic test payload:                                                                                                                                                                                                                                                                         
   
  ```json                                                                                                                                                                                                                                                                                             
  {                                                               
    "model": "Kimi-K2.6",
    "max_tokens": 8000,
    "temperature": 0,                                                                                                                                                                                                                                                                                 
    "top_p": 1,
    "messages": [                                                                                                                                                                                                                                                                                     
      {"role": "system", "content": "You are an expert presentation designer. Generate a single self-contained HTML slide with inline CSS. Canvas is 1280x720 pixels. Use absolute positioning for elements. Output ONLY the HTML starting with <html> and ending with </html>; no commentary, no
  markdown code fences."},                                                                                                                                                                                                                                                                            
      {"role": "user", "content": "Create a title slide for a 5-slide deck introducing a generic SaaS product. Include: a large title \"Meet Acme\", a subtitle \"The all-in-one collaboration platform\", and a clean modern design with a blue-to-purple gradient background. Output complete HTML
  only."}                                                                                                                                                                                                                                                                                             
    ]                                                             
  }                                                                                                                                                                                                                                                                                                   
  ```                                                                                                                                                                                                                                                                                                 
   
  The loop reproduces on most non-trivial reasoning prompts; this is just a representative one.                                                                                                                                                                                                       
                                                                  
  ### Expected vs Actual                                                                                                                                                                                                                                                                              
                                                                  
  **H100 (reference):** `reasoning_content` is fluent English from start to finish, no degenerate cycles.                                                                                                                                                                                             
                                                                  
  **GB200:** `reasoning_content` starts coherent (~150–700 chars), then collapses into a deterministic repeat. The attractor depends on which kernel variant is used:                                                                                                                                 
                                                                  
  | sgl_kernel | `--moe-runner-backend` | `--disable-cuda-graph` | Attractor |                                                                                                                                                                                                                        
  |---|---|---|---|                                               
  | 0.4.1 | triton | no | `"foss modern and "` (3-token cycle) |                                                                                                                                                                                                                                      
  | 0.4.1 | cutlass | no | `"Zoom Communications is a well-known video conferencing platform, so the design should reflect its modern, clean, and professional style"` repeated ~240× |
  | 0.4.1.post1 | cutlass | no | `"foss"` × 7833 in 8000 tokens (1-token cycle) |                                                                                                                                                                                                                     
  | 0.4.1.post1 | cutlass | yes (eager) | `"fossa"` × 3516 in 8000 tokens (1-token cycle) |                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                      
  A healthy model under greedy decode does not produce single-token repetition loops, so the model's next-token distribution is collapsing onto a delta — i.e. the logits coming out of the MoE/output stack are wrong, not just biased.                                                              
                                                                                                                                                                                                                                                                                                      
  ### Variables ruled out
                                                                                                                                                                                                                                                                                                      
  Each line is an independent A/B against the same prompt, all at `temperature=0`:

  | Variable | Tested values | Outcome on GB200 |
  |---|---|---|                                                                                                                                                                                                                                                                                       
  | Sampling temperature | 1.0, 0.0 | Both broken; T=0 reproduces deterministically |
  | `--attention-backend` | `flashinfer`, `triton`, `fa3` | All loop |                                                                                                                                                                                                                                
  | `--moe-runner-backend` | `triton`, `cutlass` | Both loop, different attractors |                                                                                                                                                                                                                  
  | MoE method (auto-selected) | `MxInt4MoE` (default with `flashinfer_trtllm`) → `WNA16MarlinMoEMethod` (with `triton`/`cutlass` runner) | Method-name selection corrected, GEMM still wrong |                                                                                                       
  | `sgl_kernel` version | 0.4.1, 0.4.1.post1 | Both broken, different attractors |                                                                                                                                                                                                                   
  | `--disable-cuda-graph` | off, on (eager) | Both broken, different attractors |                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                      
  ### Possibly related: MoE method auto-selection                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                      
  On default startup (no explicit `--moe-runner-backend`), the GB200 deployment logs:

  ```
  Using CompressedTensorsMxInt4MoE with flashinfer_trtllm backend
  ```                                                                                                                                                                                                                                                                                                 
   
  This appears incorrect for this model — the on-disk format is plain INT4 with BF16 group scales (`type: int`, `format: pack-quantized`), not MX-INT4 (block-FP / E8M0 scales). After `--moe-runner-backend triton` or `cutlass`, the log correctly reads:                                           
                                                                  
  ```                                                                                                                                                                                                                                                                                                 
  Using CompressedTensorsWNA16MarlinMoEMethod                     
  ```

  But greedy decode still loops — i.e. fixing the auto-selection alone does not fix the underlying kernel issue. Filing both observations in case they are related.                                                                                                                                   
   
  ### Hypothesis                                                                                                                                                                                                                                                                                      
                                                                  
  The W4A16 INT4 (Marlin) GEMM kernel as compiled into `sgl_kernel` for `cu130 / aarch64 / SM 100` produces numerically wrong outputs in MoE expert layers, causing the post-projection logits to collapse to a single token under greedy decoding. The bug is independent of:                        
   
  - the orchestration layer (different runner backends produce different attractors),                                                                                                                                                                                                                 
  - attention backend (all three tested produce loops),           
  - CUDA graph capture (eager mode also reproduces),
  - `sgl_kernel` patch level (`0.4.1` vs `0.4.1.post1` both broken).                                                                                                                                                                                                                                  
   
  The same `WNA16MarlinMoEMethod` on Hopper / x86_64 / cu129 wheel produces correct outputs.

**Comment by b8zhong** [comment]
@songwangnlp Hi, MXINT4 with `flashinfer_trtllm` should work, you can ignore the naming. Is there still a problem in that case? I don't think **cutlass** supports INT4A16.

**Comment by DoubleClark** [comment]
@songwangnlp  May i ask how you fix this issue? I meet the same issue no matter use --moe-runner-backend deep_gemm CompressedTensorsWNA16MarlinMoEMethod  or  --moe-runner-backend flashinfer_trtllm CompressedTensorsMxInt4MoE
